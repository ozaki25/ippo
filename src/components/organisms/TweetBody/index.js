import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import propTypes from 'prop-types';
import tweetFormat from 'src/utils/tweetFormat';
import ROUTES from 'src/constants/routes';

const StyledTweetBody = styled.div`
  margin: 0 0 5px;
  white-space: pre-wrap;
  word-wrap: break-word;
  font-size: ${({ bigText }) => (bigText ? '22px' : 'inherit')};
`;

const StyledLink = styled(Link)`
  text-decoration: none;
`;

const linkHashtag = textList => {
  return textList.map(text => {
    if (typeof text !== 'string') return text;
    const hashtagList = tweetFormat.detectHashtag(text);
    let tmp = text;
    const result = hashtagList.map((hashtag, i) => {
      const [before, ...after] = tmp.split(hashtag);
      tmp = after.join(hashtag);
      return [
        before,
        <StyledLink
          to={`${ROUTES.Tweets}?hashtag=${hashtag.replace('#', '')}`}
          key={`${hashtag}${i}`}
        >
          {hashtag}
        </StyledLink>,
      ];
    });
    return [...result, tmp].flat();
  });
};

const linkUrl = textList => {
  return textList
    .map(text => {
      if (typeof text !== 'string') return text;
      const urlList = tweetFormat.detectUrl(text);
      let tmp = text;
      const result = urlList.map((url, i) => {
        const [before, ...after] = tmp.split(url);
        tmp = after.join(url);
        return [
          before,
          <a href={url} key={`${url}${i}`}>
            {url}
          </a>,
        ];
      });
      return [...result, tmp].flat();
    })
    .flat();
};

const TweetBody = ({ text, bigText }) => {
  if (typeof text !== 'string')
    return <StyledTweetBody bigText={bigText ? 1 : 0}>{text}</StyledTweetBody>;
  const tmpTweet = linkUrl([text]);
  const tmp2Tweet = linkHashtag(tmpTweet);
  return <StyledTweetBody bigText={bigText ? 1 : 0}>{tmp2Tweet}</StyledTweetBody>;
};

TweetBody.displayName = 'TweetBody';

TweetBody.propTypes = {
  text: propTypes.node.isRequired,
  bigText: propTypes.bool,
};

TweetBody.defaultProps = {
  bigText: false,
};

export default TweetBody;
