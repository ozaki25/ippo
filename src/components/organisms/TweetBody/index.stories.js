import React from 'react';
import { storiesOf } from '@storybook/react';
import { text } from '@storybook/addon-knobs';
import TweetBody from '.';

const stories = storiesOf('organisms/TweetBody', module);

const props = {
  text: text(
    'text',
    `Storybookいいところ
  - どんなコンポーネントが存在してるか可視化できる
  - コンポーネントの扱い方が分かる
  - UIの修正デバッグが楽
  #ThinkAtomicDesign`,
  ),
};

stories.add('通常パターン', () => <TweetBody {...props} />);

stories.add('文字サイズ大きめ', () => <TweetBody {...props} bigText />);
