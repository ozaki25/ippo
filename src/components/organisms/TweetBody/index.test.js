import React from 'react';
import { snapshot } from 'src/test/helpers';
import TweetBody from '.';

const props = {
  text: `Storybookいいところ
  - どんなコンポーネントが存在してるか可視化できる
  - コンポーネントの扱い方が分かる
  - UIの修正デバッグが楽
  #ThinkAtomicDesign`,
};

snapshot('TweetBody/nomal', <TweetBody {...props} />);

snapshot('TweetBody/big', <TweetBody {...props} bigText />);
