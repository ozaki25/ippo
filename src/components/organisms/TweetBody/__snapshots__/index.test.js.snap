// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot: TweetBody/big 1`] = `
.c0 {
  margin: 0 0 5px;
  white-space: pre-wrap;
  word-wrap: break-word;
  font-size: 22px;
}

.c1 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MockRouter
    createHref={[Function]}
    location=""
    params={Object {}}
    path=""
    push={[Function]}
    replace={[Function]}
    url=""
  >
    <TweetBody
      bigText={true}
      text="Storybookいいところ
  - どんなコンポーネントが存在してるか可視化できる
  - コンポーネントの扱い方が分かる
  - UIの修正デバッグが楽
  #ThinkAtomicDesign"
    >
      <styled.div
        bigText={1}
      >
        <StyledComponent
          bigText={1}
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": false,
                "lastClassName": "c0",
                "rules": Array [
                  "
  margin: 0 0 5px;
  white-space: pre-wrap;
  word-wrap: break-word;
  font-size: ",
                  [Function],
                  ";
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="c0"
          >
            Storybookいいところ
  - どんなコンポーネントが存在してるか可視化できる
  - コンポーネントの扱い方が分かる
  - UIの修正デバッグが楽
  
            <Styled(Link)
              key="#ThinkAtomicDesign0"
              to="/tweets?hashtag=ThinkAtomicDesign"
            >
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": true,
                      "lastClassName": "c1",
                      "rules": Array [
                        "
  text-decoration: none;
",
                      ],
                    },
                    "displayName": "Styled(Link)",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": [Function],
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                    Symbol(Symbol.hasInstance): [Function],
                  }
                }
                forwardedRef={null}
                to="/tweets?hashtag=ThinkAtomicDesign"
              >
                <Link
                  className="c1"
                  replace={false}
                  to="/tweets?hashtag=ThinkAtomicDesign"
                >
                  <a
                    className="c1"
                    onClick={[Function]}
                  >
                    #ThinkAtomicDesign
                  </a>
                </Link>
              </StyledComponent>
            </Styled(Link)>
          </div>
        </StyledComponent>
      </styled.div>
    </TweetBody>
  </MockRouter>
</Provider>
`;

exports[`should match snapshot: TweetBody/nomal 1`] = `
.c0 {
  margin: 0 0 5px;
  white-space: pre-wrap;
  word-wrap: break-word;
  font-size: inherit;
}

.c1 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MockRouter
    createHref={[Function]}
    location=""
    params={Object {}}
    path=""
    push={[Function]}
    replace={[Function]}
    url=""
  >
    <TweetBody
      bigText={false}
      text="Storybookいいところ
  - どんなコンポーネントが存在してるか可視化できる
  - コンポーネントの扱い方が分かる
  - UIの修正デバッグが楽
  #ThinkAtomicDesign"
    >
      <styled.div
        bigText={0}
      >
        <StyledComponent
          bigText={0}
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": false,
                "lastClassName": "c0",
                "rules": Array [
                  "
  margin: 0 0 5px;
  white-space: pre-wrap;
  word-wrap: break-word;
  font-size: ",
                  [Function],
                  ";
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="c0"
          >
            Storybookいいところ
  - どんなコンポーネントが存在してるか可視化できる
  - コンポーネントの扱い方が分かる
  - UIの修正デバッグが楽
  
            <Styled(Link)
              key="#ThinkAtomicDesign0"
              to="/tweets?hashtag=ThinkAtomicDesign"
            >
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": true,
                      "lastClassName": "c1",
                      "rules": Array [
                        "
  text-decoration: none;
",
                      ],
                    },
                    "displayName": "Styled(Link)",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": [Function],
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                    Symbol(Symbol.hasInstance): [Function],
                  }
                }
                forwardedRef={null}
                to="/tweets?hashtag=ThinkAtomicDesign"
              >
                <Link
                  className="c1"
                  replace={false}
                  to="/tweets?hashtag=ThinkAtomicDesign"
                >
                  <a
                    className="c1"
                    onClick={[Function]}
                  >
                    #ThinkAtomicDesign
                  </a>
                </Link>
              </StyledComponent>
            </Styled(Link)>
          </div>
        </StyledComponent>
      </styled.div>
    </TweetBody>
  </MockRouter>
</Provider>
`;
